Scheduling requirements

jkey-eng opened the pull request:
TBD - Adding Draft Pull Request to prevent rework

Currently working to understand how to handle the overlap of the pre-existing software requirements `thread_sheduling`.

It appears we should ensure consistency between the documentation sections (Ex:`thread_scheduling` -> `Scheduling`).

jkey-eng committed:
software: Initial Perplexity Reqts

jkey-eng committed:
Include System Reqts

jkey-eng committed:
software: Review docs & Update generated requirements

jkey-eng committed:
Scheduling: Change Subject to Zephyr RTOS

peter-mitsis reviewed (commented):
We are on the right track, but some items need some rewording. I've tried to offer some suggestions in some places, but some of those still seem a little off to me.

nashif commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,212 @@
+COMPONENT: Scheduling
+TITLE: Ready Queue Scalability
+STATEMENT: >>>
+The scheduler shall support configurable ready queue implementations (DUMB/RBTREE/MULTIQ) with O(1) or O(log n) time complexity as documented.
```
drop the priority queue names from the requirement, those are implementation details. DUMB btw is not called DUMB anymore.

jkey-eng replied:
Reworded to: "The Zephyr RTOS shall support ready queue implementations that are configurable at build time."

Removed all implementation names per nashif's feedback below. Added "at build time" to clarify it's not runtime-configurable.

jkey-eng committed:
Scheduling: Reorganize intermingled thead requirements

Moved the requirements in thread_scheduling.sdoc to submodule documents
that align with their needs.

jkey-eng committed:
scheduling: address review feedback from peter-mitsis and nashif

Reword requirements based on PR #85 review comments:
- 001: clarify "permitted to execute on the current CPU"
- 002: EDF is a tie-breaker between equal static priority threads
- 003: drop implementation names, clarify build-time configurability
- 004: expand reschedule trigger list (priority change, scheduler unlock)
- 005: reword to avoid implying non-atomic rescheduling elsewhere
- 006: "negative static priority", add Meta-IRQ exception
- 007: "preempted" not "interrupted", "at a reschedule point" not "immediately"
- 008: yield applies to all threads, not just cooperative
- 009: "not more than" configurable ticks, gate on time slicing enabled
- 011: drop ±1 tick claim, add premature wakeup path
- 012: simplify busy wait to behavioral statement
- Fix duplicate UID (two 23-011s), renumber Thread Waking to 23-013
- Add 016: scheduler cross-CPU mutual exclusion (peter-mitsis)

Signed-off-by: Joel Key <Joel@keyengineeringllc.com>
Assisted-by: Claude:claude-opus-4.6

jkey-eng commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,212 @@
+COMPONENT: Scheduling
+TITLE: Priority-Based Selection
+STATEMENT: >>>
+The scheduler shall select the highest priority ready thread as the current thread from the ready queue.
```
Reworded to: "The Zephyr RTOS shall select the highest-priority ready thread permitted to execute on the current CPU as the current thread."

Dropped "from the ready queue" since running threads and the idle thread aren't in the ready queue.

jkey-eng commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,212 @@
+COMPONENT: Scheduling
+TITLE: Mandatory Context Switch Triggers
+STATEMENT: >>>
+The system shall trigger rescheduling during: thread state transitions (→SUSPENDED/WAITING), ISR returns, explicit k_yield() calls, and time slice expiration.
```
Expanded the list: "The Zephyr RTOS shall trigger rescheduling when: a thread becomes blocked or ready, a thread's static priority is modified, an ISR returns to thread context, a thread voluntarily yields, a time slice expires, or the scheduler is unlocked."

Added static priority modification and scheduler unlock. Replaced "SUSPENDED/WAITING" with "blocked or ready."

jkey-eng commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,212 @@
+COMPONENT: Scheduling
+TITLE: Atomic Rescheduling
+STATEMENT: >>>
+Rescheduling shall occur atomically at designated points to prevent partial state corruption during thread swaps.
```
Reworded to: "Rescheduling shall both occur at designated points and be atomic to prevent partial state corruption during thread swaps."

Used your suggested phrasing.

jkey-eng commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,212 @@
+COMPONENT: Scheduling
+TITLE: Cooperative Thread Non-Preemption
+STATEMENT: >>>
+Cooperative threads (negative priority) shall retain execution until explicitly yielding or entering unready state.
```
Reworded to: "The Zephyr RTOS shall ensure cooperative threads (negative static priority) retain execution until explicitly yielding or entering an unready state, except when preempted by Meta-IRQ threads."

Added "static" and the Meta-IRQ exception explicitly.

jkey-eng commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,212 @@
+COMPONENT: Scheduling
+TITLE: Preemptive Thread Interruption
+STATEMENT: >>>
+Preemptive threads (≥0 priority) shall be interrupted immediately by higher-priority ready threads.
```
Changed to "preempted" and also replaced "immediately" with "at a reschedule point" since preemption doesn't happen mid-instruction. Title updated from "Preemptive Thread Interruption" to "Preemptive Thread Preemption."

jkey-eng commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,212 @@
+COMPONENT: Scheduling
+TITLE: Cooperative Voluntary Yielding
+STATEMENT: >>>
+Cooperative threads invoking k_yield() shall allow equal/higher priority threads to execute before resuming.
```
Reworded to: "The Zephyr RTOS shall provide a mechanism for a thread to voluntarily cede control of the CPU to a thread of equal or higher priority if any exist."

Removed the cooperative-only scope, applies to all threads now. Also corrected the requirement to describe the behavioral level without naming the API function.

jkey-eng commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,212 @@
+COMPONENT: Scheduling
+TITLE: Preemptive Time Quantum
+STATEMENT: >>>
+Preemptive threads shall relinquish CPU after a configurable time slice (system ticks) if equal-priority threads are ready.
```
Reworded to: "Where time slicing is enabled, preemptive threads shall relinquish the CPU after not more than a configurable number of system ticks if there is another ready thread of equal priority."

Adopted your suggested wording and added a "Where time slicing is enabled" gate since it's a build-time configurable feature.

jkey-eng commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,212 @@
+COMPONENT: Scheduling
+TITLE: Sleep Duration Accuracy
+STATEMENT: >>>
+k_sleep() shall delay thread execution within ±1 tick of requested duration.
```
Reworded to: "The Zephyr RTOS shall provide a mechanism to delay thread execution for a specified time period, relinquishing the CPU to other ready threads. A sleeping thread may be awakened prematurely by another thread."

Incorporated your point about premature wakeup and reworded at the behavioral level. Also removed the ±1 tick precision claim, the code doesn't guarantee that.

jkey-eng commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,212 @@
+COMPONENT: Scheduling
+TITLE: Busy Wait Non-Relinquishing
+STATEMENT: >>>
+k_busy_wait() shall not yield CPU or trigger rescheduling during delay.
```
Reworded to: "The Zephyr RTOS shall provide a mechanism to delay thread execution for a specified duration without relinquishing the CPU or triggering rescheduling."

Kept at behavioral level without naming the API function.

jkey-eng commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,212 @@
+COMPONENT: Scheduling
+TITLE: Cross-CPU Atomicity
+STATEMENT: >>>
+Spinlocks shall enforce cross-CPU mutual exclusion for scheduler data structures.
```
Added as ZEP-SRS-23-016 with your exact wording.

jkey-eng commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -30,7 +30,7 @@ TYPE: Functional
 TITLE: Deadline Enforcement (EDF)
 STATEMENT: >>>
-When EDF scheduling is enabled, the system shall prioritize threads with equal static priority based on earliest deadline.
+When earliest-deadline-first (EDF) scheduling is enabled, the scheduler shall prioritize threads with equal static priority based on earliest deadline.
```
Reworded to: "When earliest-deadline-first scheduling (EDF) is enabled, the scheduler shall only consider the deadline when prioritizing between threads of equal static priority. The thread with the earliest deadline shall be considered to have the higher priority."

Adopted your suggested wording.

jkey-eng commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -40,13 +40,13 @@ TYPE: Non-Functional
 TITLE: Ready Queue Scalability
 STATEMENT: >>>
-The scheduler shall support configurable ready queue implementations (DUMB/RBTREE/MULTIQ) with O(1) or O(log n) time complexity as documented.
+The scheduler shall support configurable ready queue implementations.
```
Reworded to: "The Zephyr RTOS shall support ready queue implementations that are configurable at build time."

Removed all implementation names per nashif's feedback below. Added "at build time" to clarify it's not runtime-configurable.

---

Note: This requirement ended up having two conversation items around it. I added the feedback in both. Planning to test fixup commits moving forward to see if that'll improve the Github review feedback tracking.

jkey-eng commented:
Addressed all review comments from @peter-mitsis and @nashif in the latest commit — replies are inline on each thread.

This is also a test of the fixup commit workflow for handling review feedback. Since the earlier reorganization commit shifted line numbers, the original review comments are showing as outdated in the GitHub UI. Going forward, new review comments on the latest commit should anchor correctly.

@tobiaskaestner - if you have a chance, try leaving a review comment on the latest commit so we can confirm the comment anchoring works as expected with the fixup approach.

Moving this from draft to ready for review.

jkey-eng committed:
scheduling: add coverage gaps from kernel docs

- Reword 010: clarify scheduler lock behavior including persistence
  across unready/ready transitions
- Add 017: time slice priority exemption — threads above a configurable
  threshold are exempt from preemptive time slicing

Assisted-by: Claude:claude-opus-4.6
Signed-off-by: Joel Key <Joel@keyengineeringllc.com>

peter-mitsis reviewed (approved):
This looks good to me (setting aside the merge conflicts github is reporting).

anobli commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,224 @@
+STATEMENT: >>>
+The Zephyr RTOS shall select the highest-priority ready thread permitted to execute on the current CPU as the current thread.
+<<<
+
```
We miss the parent:
```suggestion
RELATIONS:
- TYPE: Parent
  VALUE: ZEP-SYRS-11
- TYPE: Parent
  VALUE: ZEP-SYRS-12
 
```

anobli commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,224 @@
+COMPONENT: Thread Scheduling
+TITLE: Priority-Based Selection
+STATEMENT: >>>
+The Zephyr RTOS shall select the highest-priority ready thread permitted to execute on the current CPU as the current thread.
```
Permitted is a little bit ambiguous! Maybe you should rephrase to explain that permitted mean that the thread could run on current CPU based on the CPU affinity.

anobli commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,224 @@
+UID: ZEP-SRS-23-001
+STATUS: Active
+TYPE: Functional
+COMPONENT: Thread Scheduling
```
Thread Scheduling vs Scheduling everywhere.
I prefer Thread Scheduling as component name but anyway, we should use the same everywhere.

anobli commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,224 @@
+COMPONENT: Scheduling
+TITLE: Earliest Deadline First Scheduling
+STATEMENT: >>>
+When earliest-deadline-first (EDF) scheduling is enabled, the scheduler shall only consider the deadline when prioritizing between threads of equal static priority. The thread with the earliest deadline shall be considered to have the higher priority.
```
I think we must use where instead of when (this is an optional feature, enabled at build time).
Also, I find `the scheduler shall only consider the deadline when prioritizing between threads of equal static priority` a little confusing. We understand the intent, but, first talking about deadline and then about static priority sounds like, in the first instance, that the deadline mater more than the static priority.

```suggestion
Where earliest-deadline-first (EDF) scheduling is enabled, when two or more ready threads have equal static priority, the Zephyr RTOS shall consider the thread with the earliest deadline to have the higher priority.".```

anobli commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,224 @@
+USER_STORY: >>>
+As a Zephyr RTOS user, I want to be able to schedule threads by earliest deadline first.
+<<<
+
```
We miss the parent:
```suggestion
RELATIONS:
- TYPE: Parent
  VALUE: ZEP-SYRS-12
 
```

anobli commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,224 @@
+COMPONENT: Scheduling
+TITLE: Ready Queue Scalability
+STATEMENT: >>>
+The Zephyr RTOS shall support ready queue implementations that are configurable at build time.
```
Alone, this requirement is confusing and doesn't give much details about what is expected, and why we need to support multiple queue algorithm. Maybe you could update the requirement to mention that need to select a queue based on memory footprint, CPU usage, number of thread to manage.
In addition, maybe we could add a requirement for each type of queue with the expected behavior for that type of queue.

anobli commented on docs/software_requirements/scheduling.sdoc:
```diff
@@ -0,0 +1,224 @@
+COMPONENT: Scheduling
+TITLE: Voluntary Yielding
+STATEMENT: >>>
+The Zephyr RTOS shall provide a mechanism for a thread to voluntarily cede control of the CPU to a thread of equal or higher priority if any exist.
```
IIRC, this one is defined in thread itself so it should probably go to thread requirements.

anobli reviewed (changes requested):
Here are some comments I had time to do today!
Further will follow in the next days!
